jinahub/encoders/image/CLIPImageEncoder/tests/unit/test_clip.py

```diff
@@ -1,86 +1,262 @@
+            Document(blob=np.ones((3, 224, 224), dtype=np.float32)),
+        ]
+    )
+    encoder_no_pre.encode(docs, {})
+    assert len(docs.get_attributes("embedding")) == 2
+    assert docs[0].embedding.shape == (_EMBEDDING_DIM,)
+    assert docs[0].embedding.dtype == np.float32
+    np.testing.assert_allclose(docs[0].embedding, docs[1].embedding)
+
+
+@pytest.mark.parametrize(
+    "path, expected_counts",
+    [["c", (("r", 0), ("c", 3), ("cc", 0))], ["cc", (("r", 0), ("c", 0), ("cc", 2))]],
+)
+def test_traversal_path(
+    path: str,
+    expected_counts: Tuple[str, int],
+    nested_docs: DocumentArray,
+    encoder: CLIPImageEncoder,
+):
+    encoder.encode(nested_docs, parameters={"traversal_paths": [path]})
+    for path_check, count in expected_counts:
+        assert (
+            len(nested_docs.traverse_flat([path_check]).get_attributes("embedding"))
+            == count
+        )
+
+
+@pytest.mark.parametrize("batch_size", [1, 2, 4, 8])
+def test_batch_size(encoder: CLIPImageEncoder, batch_size: int):
+    blob = np.ones((100, 100, 3), dtype=np.uint8)
+    docs = DocumentArray([Document(blob=blob) for _ in range(32)])
+    encoder.encode(docs, parameters={"batch_size": batch_size})
+
+    for doc in docs:
+        assert doc.embedding.shape == (_EMBEDDING_DIM,)
+
+
+@pytest.mark.parametrize("batch_size", [1, 2, 4, 8])
+def test_batch_size_no_preprocessing(encoder_no_pre: CLIPImageEncoder, batch_size: int):
+    blob = np.ones((3, 224, 224), dtype=np.uint8)
+    docs = DocumentArray([Document(blob=blob) for _ in range(32)])
+    encoder_no_pre.encode(docs, parameters={"batch_size": batch_size})
+
+    for doc in docs:
+        assert doc.embedding.shape == (_EMBEDDING_DIM,)
+
+
+def test_embeddings_quality(encoder: CLIPImageEncoder):
+    """
+    This tests that the embeddings actually "make sense".
+    We check this by making sure that the distance between the embeddings
+    of two similar images is smaller than everything else.
+    """
+
+    data_dir = Path(__file__).parent.parent / "imgs"
+    dog = Document(id="dog", blob=np.array(Image.open(data_dir / "dog.jpg")))
+    cat = Document(id="cat", blob=np.array(Image.open(data_dir / "cat.jpg")))
+    airplane = Document(
+        id="airplane", blob=np.array(Image.open(data_dir / "airplane.jpg"))
+    )
+    helicopter = Document(
+        id="helicopter", blob=np.array(Image.open(data_dir / "helicopter.jpg"))
+    )
+
+    docs = DocumentArray([dog, cat, airplane, helicopter])
+    encoder.encode(docs, {})
+
+    docs.match(docs)
+    matches = ["cat", "dog", "helicopter", "airplane"]
+    for i, doc in enumerate(docs):
+        assert doc.matches[1].id == matches[i]
+
+
+def test_openai_embed_match():
+    data_dir = Path(__file__).parent.parent / "imgs"
+    dog = Document(id="dog", blob=np.array(Image.open(data_dir / "dog.jpg")))
+    cat = Document(id="cat", blob=np.array(Image.open(data_dir / "cat.jpg")))

```

Comment:
Just remove the cat. I did some exploring, and the difference is in how HF and CLIP (torchvision transforms) implement the resize + centercrop transformations. The difference is subtle - i saved both images with only these 2 transformations and they looked visually the same, yet their pixel values were different.

Not sure why exactly this is, but in any case there is nothing that can be done about this, nor do I think it should be a major problem.